### ISSUE TITLE
Type checker can behave weirdly when using the name of the functions as a binder

### ISSUE THREAD

**Issue by eriksvedang** [carp-lang/Carp#597]
Example:

```
(defn n [xs]
  (let [xs [1 2 3]
        n &xs]
    n)) ;; The type of 'n' here is wrong, plus this should be a type error!
```

**Comment by hellerve** [comment]
Currently this seems to crash the compiler with:

```
carp: Too many variables with the same name in set (was looking for xs at line 3, column 12 in 'REPL')
CallStack (from HasCallStack):
  error, called at src/Concretize.hs:1487:22 in CarpHask-0.4.2.0-In3bsPt14LZHMVFTWVtUXQ:Concretize
  ```
  
  This actually seems like the right kind of error, it would just have to be made a “normal” error instead of a crash!

**Comment by hellerve** [comment]
Oh, that’s actually because of `xs`, not because of `n`.

**Comment by scolsen** [comment]
I think doing this correctly will be a bit tricky. Personally, I think that the code above should compile and run just fine. The compiler should have two candidate `n`'s when evaluating the final form and select the appropriate one based on typing information. Presently the error is being thrown by some borrow checking functions, so this change potentially has a wide net effect depending on how we do it.

**Comment by scolsen** [comment]
Though it's definitely a debatable point. It depends on whether or not we want the compiler to assume the user knows what she's doing or would rather it help her find problems. If the user knows what they are doing, both of these examples:

```clojure
(defn n [xs]
  (let [xs [1 2 3]
        n &xs]
    n))
```

```clojure
(defn n [xs]
  (let [xs [1 2 3]
        n &xs]
    (n [0 0 0])))
```

should work. However, it's quite possible either one of them is an unwitting mistake on the part of the author of the code, in which case it'd be good to at least have the compiler emit a warning about shadows of different types.

**Comment by scolsen** [comment]
Also, as @hellerve noted, the problem here is actually being caused by the let binding shadowing the argument.

We should likely "alpha rename" let forms anyway and have this case just be a pure shadowing of xs. Again, this should be a case where we probably warn the user like:

```
(defn n [xs]
  (let [xs [1 2 3]
        n &xs]
    n))

WARN: Argument xs is unreachable. Shadowed by let-binding `xs`.
```

**Comment by scolsen** [comment]
So, perhaps this won't be too difficult after all. I think the approach should be as follows:

1. Check for name conflicts between let forms and existing local environment bindings
2. If names conflict *and* their types are the same, alpha-rename the let case

2. will be the hard part depending on how well we already select bindings based on type-matching.

**Comment by scolsen** [comment]
Right now, the error happens because `xs` is assigned two deleters, of different types, an Array deleter and a Primitive deleter.

```
Set {unSet = fromList [(ProperDel Array.delete__int  "xs"),(PrimDel "xs")]}
```

I'm not sure exactly why this happens yet.

**Comment by scolsen** [comment]
Seems like it might ultimately stem from type constraints. The error throwing function is called before any evaluation occurs.

**Comment by eriksvedang** [comment]
OK, that's some really good hints. Have you tried printing the type constraints after they have been generated for the function? (there's a commented out trace in Infer.hs)

**Comment by scolsen** [comment]
Not yet! I'll see if I can give that a try today

**Comment by scolsen** [comment]
looks like my hunch might've been wrong--the error might be manifesting in the memory management code, since it seems like the type inference is correct:

```
Constraints for 'n':
  {Int == Int (ord OrdArrHead)} 
  {t0 == t0 (ord OrdArg)} 
  {(Ref (Array t4) t5) == (Ref (Array t4) t5) (ord OrdDefnBody)} 
  {(Array Int) == (Array Int) (ord OrdLetBind)} 
  {(Ref (Array Int) t5) == (Ref (Array Int) t5) (ord OrdLetBind)} 
  {(Ref (Array Int) t5) == (Ref (Array t4) t5) (ord OrdLetBody)} 
  {Int == Int (ord OrdArrBetween)} 
  {Int == Int (ord OrdArrBetween)} 
Mappings for 'n: Map {unMap = fromList [("t4",Int)]}
Constraints for 'n':
  {Int == Int (ord OrdArrHead)} 
  {t0 == t0 (ord OrdArg)} 
  {(Ref (Array Int) t5) == (Ref (Array Int) t5) (ord OrdDefnBody)} 
  {(Array Int) == (Array Int) (ord OrdLetBind)} 
  {(Ref (Array Int) t5) == (Ref (Array Int) t5) (ord OrdLetBind)} 
  {(Ref (Array Int) t5) == (Ref (Array Int) t5) (ord OrdLetBody)} 
  {Int == Int (ord OrdArrBetween)} 
  {Int == Int (ord OrdArrBetween)} 
Mappings for 'n: Map {unMap = fromList []}
```

I didn't include it here because the output is long, but if you look at the annotated XObj everything is type correctly. The name of the function `n` is not typed, while the `n` in the let binding and body is typed as `Ref (Array Int)`... so that's all working as its should.

The mysterious bit is the duplication of the `xs` deleter; I'm not sure where that's coming from.

**Comment by scolsen** [comment]
oh, I think the conflict arises from the fact that `xs` in the argument position is typed as `Nothing` while `xs` in the let scope is typed as `Array Int`.

So it's the shadowing of the argument name, not the function name n, that's causing the problem.

**Comment by scolsen** [comment]
so maybe that's good enough? we'll see if it breaks any tests. The approach is logically sound at least insofar as it doesn't really make sense to provide a deleter for a type variable. assuming that inference eventually concretizes the types correctly we'll eventually get the correct deleter for the concrete type

**Comment by eriksvedang** [comment]
> The approach is logically sound at least insofar as it doesn't really make sense to provide a deleter for a type variable. assuming that inference eventually concretizes the types correctly we'll eventually get the correct deleter for the concrete type

I agree. This was probably an oversight on my end. Any code with generic types can't really be checked for memory management like you say, since all the types could end up being non-managed. At least in theory. It will all get checked in the concretely typed version, later.

**Comment by scolsen** [comment]
I haven't run the tests yet, but it seems like it's set up to work like that already (where we'll eventually resolve against the concrete types) -- so I think it should be safe to use this simple fix so long as the tests pass!

I'm also going to try and make that error a little bit better (instead of spitting out the whole let form I'll have it spit out the body only)

**Comment by eriksvedang** [comment]
Awesome – great fix!

**Comment by scolsen** [comment]
uh oh, it did in fact break one testing:

```
./test/add-c.carp
You’re using a given-away value `(Test.handler state (ref x) (ref y) descr "value" =)` at line 59, column 5 in '/Users/scottolsen/dev/fresh-carp/core/Test.carp'.

You’ll have to copy the value using `@`. at /Users/scottolsen/dev/fresh-carp/core/Test.carp:58:4.

Traceback:
  (defn assert-ref-equal [state x y descr] (handler state (ref...) at /Users/scottolsen/dev/fresh-carp/core/Test.carp:58:3.
(defmodule Test (deftype State [passed Int failed Int]) (hid...) at /Users/scottolsen/dev/fresh-carp/core/Test.carp:17:1.
(load "Test.carp") at /Users/scottolsen/dev/fresh-carp/test/add-c.carp:1:1.
```

looks like it no longer likes the test module itself! this is a weird result because anything that this change affected was previously assigned a deleter anyways and likely would have triggered the same error....I think?

**Comment by scolsen** [comment]
so, I think what's happening now, is if we encounter a function definition where the *return type* is a variable, it gets no deleter, which eventually means the return type it's treated as needing to be borrowed....tricky. So we may need to differentiate between shadows and not shadows.

another approach would be: give all variables a fake deleter then implement some logic so that in the case of conflicts, fake deleters are ignored.

**Comment by scolsen** [comment]
ok! so, giving type variables a fake/dummy deleter and filtering those deleters later when we perform reference checks seems to do the trick

**Comment by scolsen** [comment]
ok so unfortunately we can't just ignore type vars since it leads to some problems. mainly what winds up happening is some forms that are initially untyped don't get their deleters in the end and are treated as owned when they arent.

It really is a shadowing problem--basically we don't update the deleters to account for the fact that the same symbol name is bound to a different scope, resulting in rare instances in the duplicate deleters that cause this bug.

there's two ways to solve it: alpha renaming, which would make sure in all places there are never ambiguities between symbols in different scopes -- but we'd need to add this as an additional preprocessing step when processing all forms.

the other way is to manage the deleter state such that we prefer the let deleters (the shadow deleter) only for the let scope. 

I've gone with the latter since it effects less of the codebase.
